lib/participant.js

```diff
@@ -296,39 +296,64 @@ class Participant extends EventEmitter {
       const publication = new RemoteTrackPublication(signaling, { log });
       self._addTrackPublication(publication);
 
-      signaling.getTrackTransceiver().then(trackReceiver => {
-        const RemoteTrack = {
-          audio: RemoteAudioTrack,
-          video: RemoteVideoTrack,
-          data: RemoteDataTrack
-        }[signaling.kind];
-
-        // NOTE(mroberts): It should never be the case that the TrackSignaling and
-        // MediaStreamTrack or DataTrackReceiver kinds disagree; however, just in
-        // case, we handle it here.
-        if (!RemoteTrack || signaling.kind !== trackReceiver.kind) {
-          return;
-        }
-
-        const track = signaling.kind === 'data'
-          ? new RemoteTrack(trackReceiver, { log, name: signaling.name })
-          : new RemoteTrack(trackReceiver, signaling.isEnabled, { log, name: signaling.name });
+      let isSubscribed = signaling.isSubscribed;
+      if (isSubscribed) {
+        trackSignalingSubscribed(signaling);
+      }
 
-        self._addTrack(track, publication);
+      signaling.on('updated', () => {
+        if (isSubscribed !== signaling.isSubscribed) {
+          isSubscribed = signaling.isSubscribed;
+          if (isSubscribed) {
+            trackSignalingSubscribed(signaling);
+            return;
+          }
+          trackSignalingUnsubscribed(signaling);
+        }
       });
     }
 
     function trackSignalingRemoved(signaling) {
-      signaling.getTrackTransceiver().then(() => {
-        const publication = self.trackPublications.get(signaling.sid);
-        const track = self.tracks.get(signaling.id);
-        if (track) {
-          self._removeTrack(track, publication);
-        }
-        if (publication) {
-          self._removeTrackPublication(publication);
-        }
-      });
+      if (signaling.isSubscribed) {
+        signaling.setTrackTransceiver(null);
+      }
+      signaling.removeAllListeners('updated');
```

Comment:
`removeAllListeners` is a risky API. If a class other than Participant set a listener on TrackSignaling, it will be removed, too. I recommend we do not use this API, and we do this differently.